---
title: "Study effect of surgery on BCC samples"
author: "Massimo Andreatta"
date: "`r format(Sys.Date(),'%e de %B, %Y')`"
output: 
  rmdformats::downcute:
    lightbox: true
    thumbnails: false
    self_contained: true
    gallery: true
    code_folding: show
  pkgdown:
    as_is: true
---

To investigate a potential causal link between ulceration and infiltrative tumor states, we performed scRNAseq at time-controlled tumor injury and 1-week longitudinal follow-up. We only included patients with non-ulcerated, non-previously biopsied tumors but high suspicion of BCC. For each patient, we collected samples from baseline (at the time of diagnostic biopsy), wounded and unwounded (at the time of following surgery, 1 week later) conditions. 

```{r, include=FALSE, fig.width=16, fig.height=12}
renv::restore()
library(Seurat)
library(ggplot2)
library(SignatuR)
library(UCell)
library(patchwork)
library(tidyr)
library(dplyr)
library(RColorBrewer)
library(viridis)
library(ggpubr)

palette <- brewer.pal(11, name="Paired")
options("Seurat.object.assay.version" = "v3")
```


# Read in data

Two patients at three timepoints: Baseline (first biopsy), 1 week after surgery on same wound, 1 week after surgery at different location

```{r}
#file_cancer <- "cache/Tumor_WH.rds"
file_cancer <- '~/Dropbox/CSI/Collaborations/Kuonen_BCC/data/rds/Tumor_WH.rds'

seu.merged <- readRDS(file_cancer)
```


Read in BCC object used to characterize the meta-programs (MPs) - from script `3-annotate_MPs.Rmd`
```{r}
map <- readRDS("_aux/BCC_seurat_NMFprograms.rds")
mps <- map@misc$MP_signatures
MPscale <- map@misc$MP_scale
```

Also include MP7-specific and MP3-specific signatures
```{r}
allgenes <- unlist(mps)
occ <- table(allgenes)
uniq <- names(occ)[occ==1]

mps.uq <- lapply(mps, function(x) {
  x[x %in% uniq]
})
names(mps.uq) <- paste0(names(mps.uq), ".uniq")
```

```{r}
seu.merged <- AddModuleScore_UCell(seu.merged, features = c(mps, mps.uq), ncores=8, name = "")
```

```{r fig.height=8}
VlnPlot(seu.merged, features=names(mps), group.by = "Sample",
        cols = palette, pt.size = 0)
```

Differences in specific MPs
```{r fig.height=8, fig.width=12}
which.programs <- c("MP7.WR", "MP3.ED","MP2", "MP4")

pll <- lapply(which.programs, function(program) {
  which.sym <- sym(program)
  
  pat1 <- subset(seu.merged, subset=Sample %in% c("Baseline1","Wound1","Unwound1"))
  pat2 <- subset(seu.merged, subset=Sample %in% c("Baseline2","Wound2","Unwound2"))
  
  library(ggridges)
  data1 <- pat1@meta.data[,c("Sample", program)]
  a <- ggplot(data1, aes(x = !!which.sym, y = Sample, fill=Sample)) +
    scale_fill_manual(values=palette) +
    geom_density_ridges() + ggtitle("Patient 1") + theme_bw() + 
    theme(axis.title.y = element_blank()) + NoLegend()
  
  data2 <- pat2@meta.data[,c("Sample", program)]
  b <- ggplot(data2, aes(x = !!which.sym, y = Sample, fill=Sample)) +
    scale_fill_manual(values=palette) +
    geom_density_ridges() + ggtitle("Patient 2") + theme_bw() +
    theme(axis.title.y = element_blank()) + NoLegend()
  
  a | b
})

wrap_plots(pll, ncol=1)
```

Differences in specific MPs (boxplot)
```{r fig.height=8, fig.width=6}
fboxplot <- function(data, program, ymax=NULL, plot.signif=TRUE) {
  
  pal.this <- c(brewer.pal(3, "Purples"), brewer.pal(3, "Blues"))
  
  which.sym <- sym(program)
  pat1 <- subset(data, subset=Sample %in% c("Baseline1","Wound1","Unwound1"))
  pat2 <- subset(data, subset=Sample %in% c("Baseline2","Wound2","Unwound2"))
  
  comparisons <- list( c("Baseline1", "Wound1"), c("Unwound1", "Wound1"))
  data1 <- pat1@meta.data[,c("Sample", program)]
  a <- ggplot(data1, aes(y = !!which.sym, x = Sample, fill=Sample)) +
    scale_fill_manual(values=pal.this[1:3]) +
    geom_boxplot(width=0.8, outlier.shape=NA) + ggtitle("Patient 1") +
    theme_bw() + theme(axis.text.x = element_blank())
  if (!is.null(ymax)) {
    a <- a + ylim(c(0,ymax[1]))
  }
  if (plot.signif) {
    a <- a + stat_compare_means(comparisons = comparisons, label="p.format")
  }
  
  comparisons <- list( c("Baseline2", "Wound2"), c("Unwound2", "Wound2"))
  data2 <- pat2@meta.data[,c("Sample", program)]
  
  b <- ggplot(data2, aes(y = !!which.sym, x = Sample, fill=Sample)) +
    scale_fill_manual(values=pal.this[4:6]) +
    geom_boxplot(width=0.8, outlier.shape=NA) + ggtitle("Patient 2") +
    theme_bw() + theme(axis.text.x = element_blank())
  if (!is.null(ymax)) {
    b <- b + ylim(c(0,ymax[2]))
  }
   if (plot.signif) {
    b <- b + stat_compare_means(comparisons = comparisons, label="p.format")
  }
  
  a | b
}
```

```{r}
pll <- list()
mp <- "MP7.WR"
fboxplot(seu.merged, program=mp, ymax=NULL, plot.signif = T)
pll[[mp]] <- fboxplot(seu.merged, program=mp, ymax=c(0.7, 0.3), plot.signif = F)

mp <- "MP2"
fboxplot(seu.merged, program=mp, ymax=NULL, plot.signif = T)
pll[[mp]] <- fboxplot(seu.merged, program=mp, ymax=c(0.7, 0.7), plot.signif = F)

mp <- "MP3.ED"
fboxplot(seu.merged, program=mp, ymax=NULL, plot.signif = T)
pll[[mp]] <- fboxplot(seu.merged, program=mp, ymax=c(0.5, 0.4), plot.signif = F)

wrap <- wrap_plots(pll, ncol=1)
ggsave("plots/wound_signature_scores_split.pdf", plot=wrap, height=7, width=5)
```

Differences in specific MPs (boxplots)
```{r fig.height=6, fig.width=9}
library(ggpubr)

which.programs <- c("MP7.WR", "MP3.ED", "MP2", "MP7.WR.uniq", "MP3.ED.uniq", "MP2.uniq")

pll <- lapply(which.programs, function(program) {
  which.sym <- sym(program)
  
  data <- seu.merged@meta.data[,c("Sample", program)]
  data$Sample <- factor(data$Sample, levels=c("Baseline1","Unwound1","Wound1","Baseline2","Unwound2","Wound2"))
  
  data$Patient <- "Patient1"
  data$Patient[data$Sample %in% c("Baseline2","Wound2","Unwound2")] <- "Patient2"
  
  pal.this <- c(brewer.pal(3, "Purples"), brewer.pal(3, "Blues"))
  
  a <- ggplot(data, aes(y = !!which.sym, x = Patient, fill=Sample)) +
    scale_fill_manual(values=pal.this) +
    geom_boxplot(width=0.8, color="black", outlier.shape=NA) + ggtitle(program) +
    ylim(0,0.6) +
    theme_bw() + theme(axis.text.x = element_blank())
  
})

wrap_plots(pll, ncol=3)
ggsave("plots/wound_signature_scores.pdf", height=6, width=9)
```



Individual marker genes
```{r fig.width=14, fig.height=4}
genes <- c("LGR5","LRP8","LAMB1","TMTC1","MEGF6",
           "LGALS7B","CLDN4","CALML5","DMKN","KRT75",
           "PHLDA2","CHI3L1","TYMP","IFITM3","TIMP1")

#DotPlot(seu.merged, features=genes, group.by = "Sample", cols="RdBu", scale=F, col.max=3) +
#  theme(axis.text.x = element_text(angle = 45, hjust=1))

pat1 <- subset(seu.merged, subset=Sample %in% c("Baseline1","Unwound1","Wound1"))
p1 <- DotPlot(pat1, features=genes, group.by = "Sample", cols="RdBu", scale=T, col.max=3) +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) + coord_flip()

pat2 <- subset(seu.merged, subset=Sample %in% c("Baseline2","Unwound2","Wound2"))
p2 <- DotPlot(pat2, features=genes, group.by = "Sample", cols="RdBu", scale=T, col.max=3) +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) + coord_flip()

p1 | p2

ggsave("plots/wound_genepanel_dotplot.pdf", height=4.5, width=10)
```

```{r fig.width=14, fig.height=4}
#function 'scale' with pseudo-count
scale.ps <- function(x, pseudo=0) {
  x <- as.numeric(x)
  mu <- mean(x, na.rm = T)
  sd <- sd(x, na.rm = T) + pseudo
  (x - mu)/sd
}

DotPlot.pseudo <- function(obj, features, assay="RNA", group.by=NULL, pseudo=0) {
  
  if (!is.null(group.by)) {
    seu.list <- SplitObject(obj, split.by = group.by)
  } else {
    seu.list <- list("Sample", obj)
  }
  
  #Percent expressed
  pct.exp <- lapply(seu.list, function(x) {
    data <- LayerData(x, assay=assay, layer = "data")[features,]
    apply(data, 1, function(g) {
      sum(g>0)/length(g)
    })
  })
  df.exp <- as.data.frame(pct.exp)
  df.exp$Gene <- rownames(df.exp)
  a <- reshape2::melt(df.exp)
  colnames(a) <- c("Gene","Sample","Pct.exp")
  
  #Average rescaled expression
  means <- lapply(seu.list, function(x) {
    data <- LayerData(x, assay=assay, layer = "data")[features,]
    apply(data, 1, mean)
  })
  df <- as.data.frame(means)
  
  #Scale expression
  df.scaled <- t(apply(df, 1, function(x) {
    scale.ps(x, pseudo=pseudo)
  }))
  colnames(df.scaled) <- colnames(df)
  
  df.scaled <- as.data.frame(df.scaled)
  df.scaled $Gene <- rownames(df.scaled)
  b <- reshape2::melt(df.scaled)
  colnames(b) <- c("Gene","Sample","Avg.exp")
  
  #Combine data and plot
  df.all <- merge(a, b)
  df.all$Gene <- factor(df.all$Gene, levels=features)
  
  cols <- RColorBrewer::brewer.pal("RdBu", n = 10)
  p <- ggplot(data = df.all, mapping = aes(x=Gene, y=Sample)) +
    geom_point(mapping = aes(size = Pct.exp, color = Avg.exp)) +
    scale_color_gradient2(low = "#3933ff", high = "#cb0000", mid = "#ffffff") +
    theme_cowplot() +
    theme(axis.text.x = element_text(angle = 45, hjust=1))
  return(p)
}
```

```{r}
library(cowplot)
genes <- c("LGR5","LRP8","LAMB1","TMTC1","MEGF6",
           "LGALS7B","CLDN4","CALML5","DMKN","KRT75",
           "PHLDA2","CHI3L1","TYMP","IFITM3","TIMP1")

pat1 <- subset(seu.merged, subset=Sample %in% c("Baseline1","Unwound1","Wound1"))
pat2 <- subset(seu.merged, subset=Sample %in% c("Baseline2","Unwound2","Wound2"))

#Normalize with pseudo
a <- DotPlot.pseudo(pat1, features = genes, pseudo = 0, group.by = "Sample") + coord_flip()
b <- DotPlot.pseudo(pat2, features = genes, pseudo = 0, group.by = "Sample") + coord_flip()
a | b

a <- DotPlot.pseudo(pat1, features = genes, pseudo = 0.5, group.by = "Sample") + coord_flip()
b <- DotPlot.pseudo(pat2, features = genes, pseudo = 0.5, group.by = "Sample") + coord_flip()
a | b

ggsave("plots/wound_genepanel_dotplot_pseudocount.pdf", height=4.5, width=8)
```

Differential expression between conditions
```{r fig.height=6, fig.width=12}
library(EnhancedVolcano)
library(SignatuR)
genes.highlight <- c("LGR5","LRP8","LAMB1","TMTC1","MEGF6",
           "LGALS7B","CLDN4","CALML5","DMKN","KRT75",
           "PHLDA2","CHI3L1","TYMP","IFITM3","TIMP1")

genes.highlight <- unname(unlist(mps[c("MP3.ED","MP2","MP7.WR")]))

sigs <- GetSignature(SignatuR$Hs)
bl <- sigs[c("Pseudogenes","Non-coding","Mito","Ribo")]
bl <- unlist(bl)

genes2test <- setdiff(rownames(seu.merged), bl)

seu.merged$Sample2g <- "Baseline_or_Unwound1"
seu.merged$Sample2g[seu.merged$Sample == "Wound1"] <- "Wound1"
seu.merged$Sample2g[seu.merged$Sample %in% c("Baseline2","Unwound2")] <- "Baseline_or_Unwound2"
seu.merged$Sample2g[seu.merged$Sample == "Wound2"] <- "Wound2"
#table(seu.merged$Sample, seu.merged$Sample2g)
Idents(seu.merged) <- "Sample2g"

#Patient1
degs <- FindMarkers(seu.merged, features=genes2test,
                    ident.1 = "Wound1", ident.2 = "Baseline_or_Unwound1", min.pct = 0.1,
                    logfc.threshold = 0.01,
                    max.cells.per.ident = 2000)

pcex <- rep(0.8, nrow(degs))
pcex[rownames(degs) %in% genes.highlight] <- 3
palpha <- rep(0.3, nrow(degs))
palpha[rownames(degs) %in% genes.highlight] <- 1

labcol <- rep("gray50", nrow(degs))
#labcol[rownames(degs) %in% c("LGR5","LRP8","LAMB1","TMTC1","MEGF6")] <- "blue2"
#labcol[rownames(degs) %in% c("LGALS7B","CLDN4","CALML5","DMKN","KRT75")] <- "forestgreen"
#labcol[rownames(degs) %in% c("PHLDA2","CHI3L1","TYMP","IFITM3","TIMP1")] <- "red2"

labcol[rownames(degs) %in% mps$MP3.ED] <- "#0ecddd"
labcol[rownames(degs) %in% mps$MP2] <- "#77dd0e"
labcol[rownames(degs) %in% mps$MP7.WR] <- "#990dd8"

names(labcol) <- rownames(degs)

p1 <- EnhancedVolcano(degs, lab=rownames(degs), selectLab = genes.highlight, x = "avg_log2FC", y = "p_val_adj",
                      pCutoff = 1e-09,
                      FCcutoff = 0.5, labSize = 5, legendPosition = "none", drawConnectors = T, pointSize = pcex,
                      colCustom = labcol, colAlpha = palpha,
                      title = "Wound1 vs. Baseline1/Unwound1", subtitle = NULL)


#Patient2
degs <- FindMarkers(seu.merged, features=genes2test,
                    ident.1 = "Wound2", ident.2 = "Baseline_or_Unwound2",
                    logfc.threshold = 0.01,
                    min.pct = 0.1, max.cells.per.ident = 2000)

pcex <- rep(0.8, nrow(degs))
pcex[rownames(degs) %in% genes.highlight] <- 3

palpha <- rep(0.3, nrow(degs))
palpha[rownames(degs) %in% genes.highlight] <- 1

labcol <- rep("gray50", nrow(degs))
#labcol[rownames(degs) %in% c("LGR5","LRP8","LAMB1","TMTC1","MEGF6")] <- "blue2"
#labcol[rownames(degs) %in% c("LGALS7B","CLDN4","CALML5","DMKN","KRT75")] <- "forestgreen"
#labcol[rownames(degs) %in% c("PHLDA2","CHI3L1","TYMP","IFITM3","TIMP1")] <- "red2"

labcol[rownames(degs) %in% mps$MP3.ED] <- "#0ecddd"
labcol[rownames(degs) %in% mps$MP2] <- "blue2"
labcol[rownames(degs) %in% mps$MP7.WR] <- "red2"

names(labcol) <- rownames(degs)

p2 <- EnhancedVolcano(degs, lab=rownames(degs), selectLab = genes.highlight, x = "avg_log2FC", y = "p_val_adj", pCutoff = 1e-09,
    FCcutoff = 0.5, labSize = 5, legendPosition = "none", drawConnectors = T, pointSize = pcex,
    colCustom = labcol, colAlpha = palpha,
    title = "Wound2 vs. Baseline2/Unwound2", subtitle = NULL)

p1 | p2

```


Differential expression between conditions
```{r fig.height=6, fig.width=12}
library(EnhancedVolcano)
library(SignatuR)
genes.highlight <- unname(unlist(mps[c("MP2","MP7.WR")]))

pval_cutoff <- 10^(-8)
min.pct <- 0.1

sigs <- GetSignature(SignatuR$Hs)
bl <- sigs[c("Pseudogenes","Non-coding","Mito","Ribo")]
bl <- unlist(bl)

genes2test <- setdiff(rownames(seu.merged), bl)

seu.merged$Sample2g <- "Baseline_or_Unwound1"
seu.merged$Sample2g[seu.merged$Sample == "Wound1"] <- "Wound1"
seu.merged$Sample2g[seu.merged$Sample %in% c("Baseline2","Unwound2")] <- "Baseline_or_Unwound2"
seu.merged$Sample2g[seu.merged$Sample == "Wound2"] <- "Wound2"
Idents(seu.merged) <- "Sample2g"

#Patient1
degs <- FindMarkers(seu.merged, features=genes2test,
                    ident.1 = "Wound1", ident.2 = "Baseline_or_Unwound1", min.pct = min.pct,
                    logfc.threshold = 0.01, 
                    max.cells.per.ident = 2000)

#Trim extreme p-values
min(degs$p_val_adj)

pcex <- rep(0.8, nrow(degs))
pcex[rownames(degs) %in% genes.highlight] <- 3
palpha <- rep(0.3, nrow(degs))
palpha[rownames(degs) %in% genes.highlight] <- 0.8

labcol <- rep("gray50", nrow(degs))

labcol[rownames(degs) %in% mps$MP2] <- "blue2"
labcol[rownames(degs) %in% mps$MP7.WR] <- "red2"
names(labcol) <- rownames(degs)

#Genes to highlight
degs.signif <- degs[degs$p_val_adj <= pval_cutoff,]
genes.highlight <- genes.highlight[genes.highlight %in% rownames(degs.signif)] 


p1 <- EnhancedVolcano(degs, lab=rownames(degs), selectLab = genes.highlight, x = "avg_log2FC", y = "p_val_adj",
                      pCutoff = pval_cutoff,
                      FCcutoff = 0.5, labSize = 5, legendPosition = "none", drawConnectors = T, pointSize = pcex,
                      colCustom = labcol, colAlpha = palpha, max.overlaps=Inf,
                      title = "Wound1 vs. Baseline1/Unwound1", subtitle = NULL)


#Patient2
genes.highlight <- unname(unlist(mps[c("MP2","MP7.WR")]))

degs <- FindMarkers(seu.merged, features=genes2test,
                    ident.1 = "Wound2", ident.2 = "Baseline_or_Unwound2",
                    logfc.threshold = 0.01,
                    min.pct = 0.1, max.cells.per.ident = 2000)

#Trim extreme p-values
min(degs$p_val_adj)
degs$p_val_adj[degs$p_val_adj < 10^(-60)] <- 10^(-60)


pcex <- rep(0.8, nrow(degs))
pcex[rownames(degs) %in% genes.highlight] <- 3

palpha <- rep(0.3, nrow(degs))
palpha[rownames(degs) %in% genes.highlight] <- 0.8

labcol <- rep("gray50", nrow(degs))
labcol[rownames(degs) %in% mps$MP2] <- "blue2"
labcol[rownames(degs) %in% mps$MP7.WR] <- "red2"

names(labcol) <- rownames(degs)

#Genes to highlight
degs.signif <- degs[degs$p_val_adj <= pval_cutoff,]
genes.highlight <- genes.highlight[genes.highlight %in% rownames(degs.signif)]


p2 <- EnhancedVolcano(degs, lab=rownames(degs), selectLab = genes.highlight, x = "avg_log2FC", y = "p_val_adj",
                      pCutoff = pval_cutoff,
                      FCcutoff = 0.5, labSize = 5, legendPosition = "none", drawConnectors = T, pointSize = pcex,
                      colCustom = labcol, colAlpha = palpha, max.overlaps=Inf,
                      title = "Wound2 vs. Baseline2/Unwound2", subtitle = NULL)

p1 | p2

ggsave("plots/wounded_vs_baseline_volcano.pdf", height=6, width=12)
```



```{r}
mat <- as.matrix(seu.merged@meta.data[,names(mps)])

#scale using centering and scaling factor determined in discovery dataset
mat.scaled <- mat
for (r in 1:nrow(mat)) {
  mat.scaled[r,] <- (mat[r,] - MPscale$center) / MPscale$sd 
}

colnames(mat.scaled) <- paste0("MPdim_",seq(1, ncol(mat.scaled)))

#New dim reduction
seu.merged@reductions[["MPs"]] <- new("DimReduc",
                                         cell.embeddings = mat.scaled,
                                         assay.used = "RNA",
                                         key = "MPdim_",
                                         global = FALSE)
set.seed(1234)
seu.merged <- RunUMAP(seu.merged, reduction="MPs",
                      dims=1:length(seu.merged@reductions[["MPs"]]),
               metric = "euclidean", reduction.name = "umap_MP")
```

```{r}
DimPlot(seu.merged, group.by = "Sample", reduction="umap_MP") + theme(aspect.ratio = 1)
```

```{r fig.height=10}
FeaturePlot(seu.merged, features = names(mps), ncol=3, pt.size = 4, reduction="umap_MP", raster = T) &
  scale_color_viridis(option="B") &
   theme(aspect.ratio = 1, axis.text=element_blank(), axis.ticks=element_blank(), axis.title = element_blank())
```

Analyze samples separately, apply knn-smoothing?
```{r}
seu.list <- SplitObject(seu.merged, split.by = "Sample")

#Smooth UCell scores
seu.list <- lapply(seu.list, function(this) {
  this <- this |> FindVariableFeatures(nfeatures = 1000) |>
    ScaleData() |> RunPCA(npcs=30) |> RunUMAP(dims=1:30, reduction.name = "umap")
  
  this <- SmoothKNN(this, reduction = "pca", signature.names = names(mps), k = 10)
  this
})
```

```{r}
s <- "Wound1"
this <- seu.list[[s]]
FeaturePlot(this, reduction="umap",
            features=c("nFeature_RNA","nCount_RNA","MP7.WR_kNN","MP3.ED_kNN")) &
  theme(aspect.ratio = 1)
```

```{r fig.height=8}
seu.merged <- Reduce(f=merge, x=seu.list)

seu.merged$Sample <- factor(seu.merged$Sample,
                            levels=c("Baseline1","Unwound1","Wound1",
                                     "Baseline2","Unwound2","Wound2"))

VlnPlot(seu.merged, features=paste0(names(mps),"_kNN"), group.by = "Sample",
        cols = palette, pt.size = 0)
```

Differences in specific MPs
```{r fig.height=8, fig.width=12}
which.programs <- c("MP7.WR_kNN", "MP3.ED_kNN","MP2_kNN")

pll <- lapply(which.programs, function(program) {
  which.sym <- sym(program)
  
  pat1 <- subset(seu.merged, subset=Sample %in% c("Baseline1","Wound1","Unwound1"))
  pat2 <- subset(seu.merged, subset=Sample %in% c("Baseline2","Wound2","Unwound2"))
  
  library(ggridges)
  data1 <- pat1@meta.data[,c("Sample", program)]
  a <- ggplot(data1, aes(x = !!which.sym, y = Sample, fill=Sample)) +
    scale_fill_manual(values=palette) +
    geom_density_ridges() + ggtitle("Patient 1") + theme_bw() + 
    theme(axis.title.y = element_blank()) + NoLegend()
  
  data2 <- pat2@meta.data[,c("Sample", program)]
  b <- ggplot(data2, aes(x = !!which.sym, y = Sample, fill=Sample)) +
    scale_fill_manual(values=palette) +
    geom_density_ridges() + ggtitle("Patient 2") + theme_bw() +
    theme(axis.title.y = element_blank()) + NoLegend()
  
  a | b
})

wrap_plots(pll, ncol=1)
```

Differences in specific MPs (violin + boxplot)
```{r fig.height=3, fig.width=9}
which.programs <- c("MP7.WR_kNN", "MP3.ED_kNN","MP2_kNN")

pll <- lapply(which.programs, function(program) {
  which.sym <- sym(program)
  
  data <- seu.merged@meta.data[,c("Sample", program)]
  data$Sample <- factor(data$Sample, levels=c("Baseline1","Unwound1","Wound1","Baseline2","Unwound2","Wound2"))
  
  data$Patient <- "Patient1"
  data$Patient[data$Sample %in% c("Baseline2","Wound2","Unwound2")] <- "Patient2"
  
  pal.this <- c(brewer.pal(3, "Purples"), brewer.pal(3, "Blues"))
  
  a <- ggplot(data, aes(y = !!which.sym, x = Patient, fill=Sample)) +
    scale_fill_manual(values=pal.this) +
    geom_boxplot(width=0.8, color="black", outlier.shape=NA) + ggtitle(program) +
    ylim(0,0.6) +
    theme_bw() + theme(axis.text.x = element_blank())

})

wrap_plots(pll, ncol=3)
ggsave("plots/wound_signature_scores_knn.pdf", height=3, width=9)
```

```{r fig.height=10, fig.width=5}
Idents(seu.merged) <- "Sample"
VlnPlot(seu.merged, features=mps$MP7.WR, pt.size=0, fill.by="ident",
        stack=T, flip = T)

found <- mps$MP7.WR[mps$MP7.WR %in% rownames(seu.merged)]
mat <- seu.merged@assays$RNA$counts[found,]

#how many of these genes have >0 counts?
hascounts <- apply(mat, 1, function(x) {sum(x>0)/length(x)})
hist(hascounts, breaks=100)

#how many of these genes are detected per cell?
cellcounts <- apply(mat, 2, function(x) {sum(x>0)})
hist(cellcounts, breaks=50)
```
